### DISCUSSION TITLE
Tuple of reference?

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#248] [General]
Just an idea that , without really using `Tuple` type. I think we might be able to return multiple reference with tuple syntax

The concept is. It does not have a real tuple object. But it force the caller to always deconstruct multiple reference returned from function sequentially

I suspect that IL could be able to do it?

```C#
struct Matrix4x4
{
    fixed float[16] _m;
    public (ref X,ref Y,ref Z) Position
    {
        get => (ref _m[12],ref _m[13],ref _m[14]);
    }

    public (ref X,ref Y,ref Z) Left
    {
        get => (ref _m[0],ref _m[4],ref _m[8]);
    }

    public (ref X,ref Y,ref Z) Forward
    {
        get => (ref _m[2],ref _m[6],ref _m[10]);
    }
}

Matrix4x4 matrix;

var position = matrix.Position;
position.X = 1; // matrix._m[12] = 1

// Not really use Tuple but actually transpiled to

(ref X_0,ref Y_0.ref Z_0) = matrix.Position;
X_0 = 1; // matrix._m[12] = 1
```

**Comment by svick** [comment]
> I suspect that IL could be able to do it

I don't see how. You can't have `ref` fields, `ref ref` parameters or more than one `ref` return value. So what is the mechanism you suggest to be used here to actually return multiple `ref`s?

**Comment by Thaina** [comment]
@svick I'm not so sure but I remember (maybe wrong) that IL could return sequence and load back in sequence

Still one way I'm quite sure it possible is returning a tuple of pointer and cast back to multiple ref var
